mockatests/drivers/test-scheduler.c

```diff
@@ -0,0 +1,1315 @@
+  scheduler_t s;
+  scheduler_initialize(&s);
+  s.max_timeout = TV_SECS(600);
+
+  const char md = SCHEDULER_POLL_TIMEOUT;
+  const int fd = 1;
+  const struct timeval to1 = { .tv_sec = 20 };
+  const struct timeval to2 = { .tv_sec = 10 };
+  fake_gettimeofday = (struct timeval){ .tv_sec = 0, .tv_usec = 0};
+  (void)scheduler_register_event(&s, md, fd, to1, &fake_event_cb, NULL);
+  (void)scheduler_register_event(&s, md, fd, to2, &fake_event_cb, NULL);
+
+  const struct timeval time_now = { .tv_sec = 2, .tv_usec = 0};
+  fake_gettimeofday = time_now;
+
+  scheduler_prepare_events(&s);
+
+  // New timeout is based on the smaller event timeout value (to2).
+  assert_int_equal(s.timeout.tv_sec, to2.tv_sec - time_now.tv_sec);
+}
+
+void
+test_scheduler_timeout_event_is_instant_if_deadline_is_now(void **state)
+{
+  // scheduler_prepare_events add TIMEOUT event timeout is zero if deadline is now
+  scheduler_t s;
+  scheduler_initialize(&s);
+  s.max_timeout = TV_SECS(600);
+
+  const char md = SCHEDULER_POLL_TIMEOUT;
+  const int fd = 1;
+  const struct timeval to = { .tv_sec = 10 };
+  fake_gettimeofday = (struct timeval){ .tv_sec = 0, .tv_usec = 0};
+  (void)scheduler_register_event(&s, md, fd, to, &fake_event_cb, NULL);
+
+  // Set the time now to the event timeout
+  fake_gettimeofday = to;
+
+  scheduler_prepare_events(&s);
+
+  // New timeout is zero because deadline has already been reached.
+  assert_int_equal(s.timeout.tv_sec, 0);
+}
+
+void
+test_scheduler_multiple_timeout_events_dont_interfere(void **state)
+{
+  // scheduler_prepare_events multiple timeout events don't clobber each other
+  scheduler_t s;
+  scheduler_initialize(&s);
+  s.max_timeout = TV_SECS(600);
+
+  const char md = SCHEDULER_POLL_TIMEOUT;
+  const int fd = 1;
+  const struct timeval to1 = { .tv_sec = 10 };
+  const struct timeval to2 = { .tv_sec = 20 };
+  fake_gettimeofday = (struct timeval){ .tv_sec = 0, .tv_usec = 0};
+  (void)scheduler_register_event(&s, md, fd, to1, &fake_event_cb, NULL);
+  (void)scheduler_register_event(&s, md, fd, to2, &fake_event_cb, NULL);
+
+  // Set the time now to the first event timeout
+  fake_gettimeofday = to1;
+
+  scheduler_prepare_events(&s);
+
+  // Event though event2 still has 10 seconds left event1 is 0 therefor timeout is 0
+  assert_int_equal(s.timeout.tv_sec, 0);
+}
+
+void
+test_scheduler_timeout_event_ignored_if_no_timeout(void **state)
+{
+  // scheduler_prepare_events add TIMEOUT event ignored if no timeout
+  scheduler_t s;
+  scheduler_initialize(&s);
+  s.max_timeout = TV_SECS(600);
+
+  const char md = SCHEDULER_POLL_TIMEOUT;
+  const int fd = 1;
+  const struct timeval to = TV_INF;
+  fake_gettimeofday = (struct timeval){ .tv_sec = 0, .tv_usec = 0};
+  (void)scheduler_register_event(&s, md, fd, to, &fake_event_cb, NULL);
+
+  scheduler_prepare_events(&s);
+
+  assert_int_equal(s.max_fd, -1);
+}
+
+void
+test_scheduler_with_no_eventS_will_timeout(void **state)
```

Comment:
looks like we have a capital S on the end of eventS ?

Reply:
Fixed in b5a2f60.